File: vm_translator/parser.py
class Parser:
    def __init__(self, filename):
        """Inicializa o parser e carrega os comandos do arquivo .vm."""
        with open(filename, "r") as file:
            self.commands = [
                line.split("//")[0].strip().split()  # Remove comentários e divide tokens
                for line in file.readlines()
                if line.strip() and not line.startswith("//")
            ]
        self.current_command = None

    def hasMoreCommands(self):
        """Retorna True se ainda há comandos a processar."""
        return bool(self.commands)

    def advance(self):
        """Lê o próximo comando e o define como corrente."""
        if self.hasMoreCommands():
            self.current_command = self.commands.pop(0)

    def commandType(self):
        """Retorna o tipo do comando."""
        if self.current_command[0] in {"add", "sub", "neg", "eq", "gt", "lt", "and", "or", "not"}:
            return "C_ARITHMETIC"
        elif self.current_command[0] == "push":
            return "C_PUSH"
        elif self.current_command[0] == "pop":
            return "C_POP"
        else:
            raise ValueError(f"Comando desconhecido: {self.current_command[0]}")

    def arg1(self):
        """Retorna o primeiro argumento do comando."""
        if self.commandType() == "C_ARITHMETIC":
            return self.current_command[0]
        return self.current_command[1]

    def arg2(self):
        """Retorna o segundo argumento (apenas para Push, Pop, Function, Call)."""
        if self.commandType() in {"C_PUSH", "C_POP"}:
            return int(self.current_command[2])
        raise ValueError(f"arg2 chamado para comando inválido: {self.current_command}")

File: codewriter GO/parser.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// Tipos de comandos possíveis
type CommandType int

const (
	Arithmetic CommandType = iota
	Push
	Pop
	Label
	Goto
	If
	Function
	Return
	Call
)

// Parser estrutura
type Parser struct {
	lines  []string
	index  int
	tokens []string
}

// Novo Parser
func NewParser(filename string) (*Parser, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := strings.Split(scanner.Text(), "//")[0] // Remove comentários
		line = strings.TrimSpace(line)                 // Remove espaços extras
		if line != "" {
			lines = append(lines, line)
		}
	}

	return &Parser{lines: lines, index: -1}, scanner.Err()
}

// Verifica se há mais comandos
func (p *Parser) HasMoreCommands() bool {
	return p.index < len(p.lines)-1
}

// Avança para o próximo comando
func (p *Parser) Advance() {
	if p.HasMoreCommands() {
		p.index++
		p.tokens = strings.Fields(p.lines[p.index])
	}
}

// Retorna o tipo do comando atual
func (p *Parser) CommandType() CommandType {
	switch p.tokens[0] {
	case "push":
		return Push
	case "pop":
		return Pop
	case "label":
		return Label
	case "goto":
		return Goto
	case "if-goto":
		return If
	case "function":
		return Function
	case "return":
		return Return
	case "call":
		return Call
	default:
		return Arithmetic
	}
}

// Retorna o primeiro argumento do comando
func (p *Parser) Arg1() string {
	if p.CommandType() == Arithmetic {
		return p.tokens[0]
	}
	return p.tokens[1]
}

// Retorna o segundo argumento (apenas para Push, Pop, Function, Call)
func (p *Parser) Arg2() int {
	if cmdType := p.CommandType(); cmdType == Push || cmdType == Pop || cmdType == Function || cmdType == Call {
		var index int
		fmt.Sscanf(p.tokens[2], "%d", &index)
		return index
	}
	return -1
}
